### PR TITLE
core: Add signal/timeout options to RunnableConfig

### PR DESCRIPTION
**Pull request by nfcampos**
- Handled by all built-in runnables
- Handled by all utility methods in base runnable, which should propagate to basically all runnables

Fixes # (issue)

### CODE REVIEW

File: langchain-core/src/runnables/base.ts [hunk 1]
```diff
@@ -430,13 +444,23 @@ export abstract class Runnable<
...
     try {
-      outputs = await func.call(
+      const promise = func.call(
         this,
         inputs,
         optionsList,
         runManagers,
         batchOptions
       );
+      outputs = optionsList?.[0]?.signal
```
Comment (jacoblee93) [inline langchain-core/src/runnables/base.ts:454]: Nice

File: langchain-core/src/runnables/remote.ts [hunk 2]
```diff
@@ -214,11 +214,12 @@ function deserialize<RunOutput>(str: string): RunOutput {
   return revive(obj);
 }
 
-function removeCallbacks(
+function removeCallbacksAndSignal(
   options?: RunnableConfig
-): Omit<RunnableConfig, "callbacks"> {
+): Omit<RunnableConfig, "callbacks" | "signal"> {
```
Comment (jacoblee93) [inline langchain-core/src/runnables/remote.ts:219]: Do we have to handle `timeout` here too?

Comment (nfcampos) [reply langchain-core/src/runnables/remote.ts:219]: No, timeout can safely be passed to JSON.stringify

File: langchain-core/src/runnables/base.ts [hunk 3]
```diff
@@ -378,7 +382,17 @@ export abstract class Runnable<
     delete config.runId;
     let output;
     try {
-      output = await func.call(this, input, config, runManager);
+      const promise = func.call(this, input, config, runManager);
+      output = options?.signal
+        ? await Promise.race([
+            promise,
+            new Promise<never>((_, reject) => {
```
Comment (jacoblee93) [inline langchain-core/src/runnables/base.ts:389]: Ah, edge case but what if someone passes an already aborted signal/there's some race condition?

Feel like there should be a check?

File: langchain-core/src/utils/signal.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+export async function raceWithSignal<T>(
+  promise: Promise<T>,
+  signal?: AbortSignal
+): Promise<T> {
+  if (signal === undefined) {
+    return promise;
+  }
+  if (signal.aborted) {
```
Comment (jacoblee93) [inline langchain-core/src/utils/signal.ts:8]: @nfcampos If we have this check here, do we need all the `signal.throwIfAborted()` calls elsewhere?

**Pull request merged**